Add a partitioner that can match the partitioning of another mesh

GiudGiud opened the pull request:
it could be made more scalable than this, but it should fit the need here

closes #30470

GiudGiud committed:
Always give a pointer to the parent app mesh for the subapp to
be able to use it for partitioning
refs #30470

GiudGiud committed:
Add a partitioner for a subapp that can copy the partitioning of the parent app
refs #30470

GiudGiud committed:
Improve comments
Add a warning on using less partitions than available
Add documentation
refs #30470

GiudGiud committed:
Implement distributed mesh support
refs #30470

GiudGiud committed:
Add tests for copy-mesh partitioner
refs #30470

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2860738/), step Docs: sync website on 36d05b1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30471/site)

This comment will be updated on new commits.

GiudGiud commented on framework/doc/content/source/partitioner/CopyMeshPartitioner.md:
```diff
@@ -0,0 +1,15 @@
+This partitioner is used to match the partitioning between two meshes.
+
+Generally, the source mesh should be partitioned with the same amount or less partitions than the target mesh is
+intended to be partitioned into. If the source mesh is partitioned with less partitions than the target mesh, some processes will not be
+assigned any element in the target mesh, leading to lesser efficiency. If the source mesh is partitioned with more partitions than the target mesh, the partitioning may only succeed if no more than the number of target partitions are used when matching the two meshes.
```
need to update this, cant actually support that case

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2861830/), step Generate coverage on 9f76841 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>6c09b2</th>
      <th colspan=4>#30471 9f7684</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.46%</td>
      <td>85.46%</td>
      <td>+0.00%</td>
      <td>86.32%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>112495</td>
      <td>112572</td>
      <td>+77</td>
      <td>82</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19138</td>
      <td>19150</td>
      <td>+12</td>
      <td>13</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30471/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30471/coverage/framework)


## Modules coverage

Inconsistent report tags were found between the head and base reports.
This can happen when reports are missing from either the head or the base.

Inconsistent tags:
  module-phase_field-recover

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30471/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 86.32% is less than the suggested 90.0%



This comment will be updated on new commits.

lindsayad commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -1129,12 +1129,13 @@ MultiApp::createApp(unsigned int i, Real start_time)
              << COLOR_DEFAULT << std::endl;
   app_params.set<unsigned int>("_multiapp_level") = _app.multiAppLevel() + 1;
   app_params.set<unsigned int>("_multiapp_number") = _first_local_app + i;
+  app_params.set<const MooseMesh *>("_master_mesh") = &_fe_problem.mesh();
```
It seems pretty wonky to me to add an additional boolean parameter, when you already have the pointer as an implicit boolean. It seems like you don't want to use a master mesh, the `_master_mesh` pointer should just remain null. Ideally this PR would not need to add another data member and API on the MooseApp

GiudGiud replied:
> It seems like you don't want to use a master mesh, the _master_mesh pointer should just remain null.

how do I get the parent app mesh for partitioning purposes only?

lindsayad replied:
It's not clear to me in the issue what your needs are. Are you oversampling in both the parent and sub-application? Just one?

lindsayad replied:
I'm trying to understand why you just want to use the parent mesh's partitioning and not the parent mesh itself

GiudGiud replied:
We could restrict the use of this object to the "clone mesh" case. I dont think. I would lift that restriction when looking at whether this is a good strategy for minimizing transfers cost

GiudGiud replied:
> I'm trying to understand why you just want to use the parent mesh's partitioning and not the parent mesh itself

In the general case, the parent and child app meshes are different. However, very often, the parent and subapp meshes overlap (both full core for example, but different discretizations for different physics). Using the same partitioning would make the Transfers do a lot less off-node communication in my opinion. 
Now, I have not shown or proven that the Transfers need this different partitioning.

GiudGiud replied:
or I could leave it as is, and use the fancy C++ keys to give access to the 3 actions to the master mesh?

lindsayad replied:
I think the situation you've outlined makes sense. If the geometries overlap but the parent and sub apps have different refinement levels or distribution of element density. So I'm good with having the boolean and mesh pointer

lindsayad commented on framework/src/partitioner/CopyMeshPartitioner.C:
```diff
@@ -0,0 +1,218 @@
+  // NOTE: in partitioner.C they manage to unpack the filled requests in the same order that
+  //       the requests are sent, thus saving the need for that last index. Possible rework..
+  std::map<processor_id_type,
+           std::vector<std::tuple<Real, Real, Real, processor_id_type, unsigned int>>>
```
You can't do a `RealVectorValue`?

GiudGiud replied:
Havent tried. Point did not work, hence the 3 Reals.
I can try

GiudGiud replied:
lemme try. I did not include that one I had the two others

lindsayad reviewed (changes requested):
I'm going to let @roystgnr lead the review on the partitioner itself

roystgnr commented on framework/src/partitioner/CopyMeshPartitioner.C:
```diff
@@ -0,0 +1,218 @@
+  std::set<unsigned int> pids_used;
+
+  // We cannot get the point locator only on some ranks of the base mesh, so we must error here
+  if (_base_mesh->comm().size() > mesh.comm().size())
```
If we want to be able to see owned source mesh elements from the corresponding target mesh's owned elements after the partitioning is done, then checking the communicator sizes isn't enough - we also need each process to be the same rank() on each communicator.

GiudGiud replied:
added an assert, not sure when this would happen

roystgnr replied:
Yeah, me either, but then I wouldn't have considered the case where the communicators are different sizes to begin with.

GiudGiud commented on framework/src/partitioner/CopyMeshPartitioner.C:
```diff
@@ -0,0 +1,218 @@
+  std::map<processor_id_type,
+           std::vector<std::tuple<Real, Real, Real, processor_id_type, unsigned int>>>
+      requested_elements;
+  const bool distributed_mesh = dynamic_cast<const DistributedMesh *>(&_base_mesh->getMesh());
```
will do

GiudGiud committed:
Address reviews

- use Point for parallel algorithm in partitioning
- better grammar
- add a check on the rank of the process in the two communicators

Co-authored-by: roystgnr <roy@stogners.org>
Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/2871934/), step Clang format on 038294f wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/30471/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/30471/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format fc26d03b7a90fb8d0acc819941d48eb1f3ea17cb`

GiudGiud commented on framework/src/partitioner/CopyMeshPartitioner.C:
```diff
@@ -0,0 +1,218 @@
+
+  // Get point locator
+  auto pl = _base_mesh->getPointLocator();
+  pl->enable_out_of_mesh_mode();
```
we can know here if the mesh is serial or not so I added an if

GiudGiud commented:
@roystgnr do you want to take another look at this? I should have addressed all your comments

GiudGiud commented:
Test failures are all unrelated.
Thanks for the reviews!
